File: src/main/java/org/usfirst/frc/team3863/wrappers/CANTalonWrapper.java
package org.usfirst.frc.team3863.wrappers;

import edu.wpi.first.wpilibj.CANTalon;

/**
 * <hr>
 * <h1>CANTalonWrapper</h1>
 * <p>A class that links, via software, two motors together.</p>
 */
public class CANTalonWrapper {

    /**
     * <hr>
     * <h1>Inner CANTalon reference</h1>
     * <p>These motors should only be accessable from with the CANTalonWrapper class.</p>
     */
    private CANTalon primary, secondary;
    double value = 0;

    /**
     * <hr>
     * <h1>CANTalonWrapper Constructor</h1>
     * <p>Ports for each motor.</p>
     */
    public CANTalonWrapper(int primary, int secondary) {
        this(primary, secondary, false);
    }

    public double getOutputValue() {
        return value;
    }

    /**
     * <hr>
     * <h1>CANTalonWrapper Constructor</h1>
     * <p>Ports for each motor, and a boolean for if they should be inverted by default.</p>
     */
    public CANTalonWrapper(int primary, int secondary, boolean invert) {
        this.primary = new CANTalon(primary);
        this.secondary = new CANTalon(secondary);
        this.secondary.changeControlMode(CANTalon.TalonControlMode.Follower);
        this.secondary.set(primary);
        setInverted(invert);
    }

    public void set(double outputValue) {
        primary.set(outputValue);
        this.value = outputValue;
    }

    public void setInverted(boolean isInverted) {
        primary.setInverted(isInverted);
        secondary.setInverted(isInverted);
    }

    public boolean getInverted() {
        return primary.getInverted();
    }

    public void set(double outputValue, byte thisValueDoesNotDoAnything) {
        set(outputValue);
    }

    public void reset() {
        primary.reset();
        secondary.reset();
        this.secondary.changeControlMode(CANTalon.TalonControlMode.Follower);
        this.secondary.set(primary.getDeviceID());
        this.value = 0;
    }


    public boolean isEnabled() {
        return primary.isEnabled();
    }


    public double getError() {
        return primary.getError();
    }

    public void setSetpoint(double setpoint) {
        primary.setSetpoint(setpoint);
        secondary.setSetpoint(setpoint);
    }

    public void reverseOutput(boolean flip) {
        primary.reverseOutput(flip);
        secondary.reverseOutput(flip);
    }

    public void reverseSensor(boolean flip) {
        primary.reverseSensor(flip);
    }

    public double get() {
        return (primary.get() + secondary.get()) / 2;
    }

    public int getEncPosition() {
        return (primary.getEncPosition() + secondary.getEncPosition()) / 2;
    }

    public void setEncPosition(int newPosition) {
        primary.setEncPosition(newPosition);
        secondary.setEncPosition(newPosition);
    }

    public int getEncVelocity() {
        return (primary.getEncVelocity() + secondary.getEncVelocity()) / 2;
    }

    public CANTalon.FeedbackDeviceStatus isSensorPresent(CANTalon.FeedbackDevice feedbackDevice) {
        return (primary.isSensorPresent(feedbackDevice).value == 1 || secondary.isSensorPresent(feedbackDevice).value == 1) ? CANTalon.FeedbackDeviceStatus.FeedbackStatusPresent : CANTalon.FeedbackDeviceStatus.FeedbackStatusUnknown;
    }


    public void configEncoderCodesPerRev(int codesPerRev) {
        primary.configEncoderCodesPerRev(codesPerRev);
        secondary.configEncoderCodesPerRev(codesPerRev);
    }


    public double getTemperature() {
        return (primary.getTemperature() + secondary.getTemperature()) / 2;
    }


    public double getOutputCurrent() {
        return (primary.getOutputCurrent() + secondary.getOutputCurrent()) / 2;
    }


    public double getBusVoltage() {
        return (primary.getBusVoltage() + secondary.getBusVoltage()) / 2;
    }


    public double getOutputVoltage() {
        return (primary.getOutputVoltage() + secondary.getOutputVoltage()) / 2;
    }


    public double getPosition() {
        return (primary.getPosition() + secondary.getPosition()) / 2;
    }


    public void setPosition(double pos) {
        primary.setPosition(pos);
        secondary.setPosition(pos);
    }


    public double getSpeed() {
        return (primary.getSpeed() + secondary.getSpeed()) / 2;
    }


    public CANTalon.TalonControlMode getControlMode() {
        return CANTalon.TalonControlMode.valueOf(primary.getControlMode().value & secondary.getControlMode().value);
    }


    public void setControlMode(int mode) {
        primary.setControlMode(mode);
    }


    public void changeControlMode(CANTalon.TalonControlMode controlMode) {
        primary.changeControlMode(controlMode);
    }


    public void setFeedbackDevice(CANTalon.FeedbackDevice device) {
        primary.setFeedbackDevice(device);
    }


    public void enable() {
        primary.enable();
        secondary.enable();
    }


    public void enableControl() {
        primary.enableControl();
        secondary.enableControl();
    }


    public boolean isControlEnabled() {
        return primary.isControlEnabled();
    }


    public void disableControl() {
        primary.disableControl();
        secondary.disableControl();
    }


    public double getP() {
        return primary.getP();
    }


    public double getI() {
        return primary.getI();
    }


    public double getD() {
        return primary.getD();
    }


    public double getF() {
        return primary.getF();
    }


    public void setP(double p) {
        primary.setP(p);
    }


    public void setI(double i) {
        primary.setI(i);
    }


    public void setD(double d) {
        primary.setD(d);
    }


    public void setF(double f) {
        primary.setF(f);
    }


    public void setIZone(int izone) {
        primary.setIZone(izone);
    }


    public double getIZone() {
        return (primary.getIZone() + secondary.getIZone()) / 2;
    }


    public void setPID(double p, double i, double d, double f, int izone, double closeLoopRampRate, int profile) {
        primary.setPID(p, i, d, f, izone, closeLoopRampRate, profile);
    }


    public void setPID(double p, double i, double d) {
        primary.setPID(p, i, d);
    }


    public void disable() {
        primary.disable();
        secondary.disable();
        this.value = 0;
    }


    public void enableBrakeMode(boolean brake) {
        primary.enableBrakeMode(brake);
        secondary.enableBrakeMode(brake);
    }

    public void stopMotors() {
        primary.stopMotor();
        secondary.stopMotor();
        this.value = 0;
    }
}

File: src/main/kotlin/org/usfirst/frc/team3863/kotlin/wrappers/CANTalonWrapper.kt
package org.usfirst.frc.team3863.kotlin.wrappers

import edu.wpi.first.wpilibj.CANTalon

/**
 * Created by Joshua Freedman on 2/11/2016.
 * Project: 2016Robot
 */
class CANTalonWrapper(i0: Int, i1: Int, invert: Boolean = false) {

    private var m0: CANTalon = CANTalon(i0)
    private var m1: CANTalon = CANTalon(i1)

    init {
        m0.inverted = invert
        m1.inverted = invert
    }

    fun set(outputValue: Double) {
        m0.set(outputValue)
        m1.set(outputValue)
    }

    var inverted: Boolean
        get() = m0.inverted && m1.inverted
        set(isInverted) {
            m0.inverted = isInverted
            m1.inverted = isInverted
        }

    operator fun set(outputValue: Double, thisValueDoesNotDoAnything: Byte) {
        m0.set(outputValue, thisValueDoesNotDoAnything)
        m1.set(outputValue, thisValueDoesNotDoAnything)
    }

    fun reset() {
        m0.reset()
        m1.reset()
    }

    val isEnabled: Boolean
        get() = m0.isEnabled || m1.isEnabled


    val error: Double
        get() = (m0.error + m1.error) / 2

    fun setSetpoint(setpoint: Double) {
        m0.setpoint = setpoint
        m1.setpoint = setpoint
    }

    fun reverseOutput(flip: Boolean) {
        m0.reverseOutput(flip)
        m1.reverseOutput(flip)
    }

    fun reverseSensor(flip: Boolean) {
        m0.reverseSensor(flip)
        m1.reverseSensor(flip)
    }

    fun get(): Double {
        return (m0.get() + m1.get()) / 2
    }

    var encPosition: Int
        get() = (m0.encPosition + m1.encPosition) / 2
        set(newPosition) {
            m0.encPosition = newPosition
            m1.encPosition = newPosition
        }

    val encVelocity: Int
        get() = (m0.encVelocity + m1.encVelocity) / 2

    fun isSensorPresent(feedbackDevice: CANTalon.FeedbackDevice): CANTalon.FeedbackDeviceStatus {
        return if (m0.isSensorPresent(feedbackDevice).value == 1 || m1.isSensorPresent(feedbackDevice).value == 1) CANTalon.FeedbackDeviceStatus.FeedbackStatusPresent else CANTalon.FeedbackDeviceStatus.FeedbackStatusUnknown
    }


    fun configEncoderCodesPerRev(codesPerRev: Int) {
        m0.configEncoderCodesPerRev(codesPerRev)
        m1.configEncoderCodesPerRev(codesPerRev)
    }


    val temperature: Double
        get() = (m0.temperature + m1.temperature) / 2


    val outputCurrent: Double
        get() = (m0.outputCurrent + m1.outputCurrent) / 2


    val busVoltage: Double
        get() = (m0.busVoltage + m1.busVoltage) / 2


    val outputVoltage: Double
        get() = (m0.outputVoltage + m1.outputVoltage) / 2


    var position: Double
        get() = (m0.position + m1.position) / 2
        set(pos) {
            m0.position = pos
            m1.position = pos
        }


    val speed: Double
        get() = (m0.speed + m1.speed) / 2


    val controlMode: CANTalon.TalonControlMode
        get() = CANTalon.TalonControlMode.valueOf(m0.controlMode.value and m1.controlMode.value)


    fun setControlMode(mode: Int) {
        m0.setControlMode(mode)
        m1.setControlMode(mode)
    }


    fun changeControlMode(controlMode: CANTalon.TalonControlMode) {
        m0.changeControlMode(controlMode)
        m1.changeControlMode(controlMode)
    }


    fun setFeedbackDevice(device: CANTalon.FeedbackDevice) {
        m0.setFeedbackDevice(device)
        m1.setFeedbackDevice(device)
    }


    fun enable() {
        m0.enable()
        m1.enable()
    }


    fun enableControl() {
        m0.enableControl()
        m1.enableControl()
    }


    val isControlEnabled: Boolean
        get() = m0.isControlEnabled || m1.isControlEnabled


    fun disableControl() {
        m0.disableControl()
        m1.disableControl()
    }


    var p: Double
        get() = (m0.p + m1.p) / 2
        set(p) {
            m0.p = p
            m1.p = p
        }


    var i: Double
        get() = (m0.i + m1.i) / 2
        set(i) {
            m0.i = i
            m1.i = i
        }


    var d: Double
        get() = (m0.d + m1.d) / 2
        set(d) {
            m0.d = d
            m1.d = d
        }


    var f: Double
        get() = (m0.f + m1.f) / 2
        set(f) {
            m0.f = f
            m1.f = f
        }


    fun setIZone(izone: Int) {
        m0.setIZone(izone)
        m1.setIZone(izone)
    }


    val iZone: Double
        get() = (m0.iZone + m1.iZone) / 2


    fun setPID(p: Double, i: Double, d: Double, f: Double, izone: Int, closeLoopRampRate: Double, profile: Int) {
        m0.setPID(p, i, d, f, izone, closeLoopRampRate, profile)
        m1.setPID(p, i, d, f, izone, closeLoopRampRate, profile)
    }


    fun setPID(p: Double, i: Double, d: Double) {
        m0.setPID(p, i, d)
        m1.setPID(p, i, d)
    }


    fun disable() {
        m0.disable()
        m1.disable()
    }


    fun enableBrakeMode(brake: Boolean) {
        m0.enableBrakeMode(brake)
        m1.enableBrakeMode(brake)
    }

    fun stopMotors() {
        m0.set(0.0)
        m1.set(0.0)
    }
}
